app/uk/gov/hmrc/agentoverseasapplicationfrontend/controllers/ApplicationController.scala

```
@@ -50,9 +50,11 @@ class ApplicationController @Inject()(
 
   def showContactDetailsForm: Action[AnyContent] = validApplicantAction.async { implicit request =>
     sessionStoreService.fetchAgentSession.map {
-      case Some(AgentSession(_, Some(contactDetails))) =>
+      case Some(AgentSession(Some(_), Some(contactDetails))) =>
         Ok(contact_details(ContactDetailsForm.form.fill(contactDetails)))
-      case _ => Ok(contact_details(ContactDetailsForm.form))
+      case Some(AgentSession(Some(_), None)) =>
+        Ok(contact_details(ContactDetailsForm.form))
+      case _ => Redirect(routes.ApplicationController.showAntiMoneyLaunderingForm())
```

This doesn't seem like a requirement at the moment. If the data is not in keystore, we should return an exception

---

its better to do redirects -  instead of throwing errors

---

By throwing error means it will be redirect to an error page or at the beginning of the journey. At this point, we dont know the amls page is the start of the journey. I think we should discuss with Charlie before redirecting to the amls page.

---

redirecting means here we are redirecting to `previous page` which is amls page ,
for example, while handling  /trading-name page if we don't find the /contact-details in the session we should redirect to /contact-details  as far as i understand 
i mean to say always redirect to its previous page

---

The problem with redirecting to previous page is, the session may be missing in the previous page too. In subscription frontend, we are dealing such issues by redirecting to the start of the journey.

---

Here is the implementation https://github.com/hmrc/agent-subscription-frontend/blob/master/app/uk/gov/hmrc/agentsubscriptionfrontend/controllers/SessionDataSupport.scala

---

yeah, if the session is missing in the previous page , it will then redirect to its previous page again , which eventually redirects to start of the journey(if every previous page's session is missing)

on the other hand, as you said, redirecting to start of the journey is better than throwing errors , also not ideal to redirect to start of the journey, if there are several pages but just one page's session is missing